test/test_main.py

```diff
@@ -1045,42 +1104,54 @@ def test_hook_and_dispatch_with_failing_hook(self):
         ]
         assert calls == expected
 
-    def test_hook_and_dispatch_but_hook_is_dispatch(self):
+    @pytest.mark.parametrize("rel,ind,path_type", [
+        (True, True, "indirect"),
+        (True, False, "direct"),
+        (False, False, "absolute_direct"),
+        (False, True, "absolute_indirect")
+    ])
+    def test_hook_and_dispatch_but_hook_is_dispatch(
+        self,
+        fake_script: FakeScript,
+        rel: bool,
+        ind: bool,
+        path_type: str,
+    ):
         event = EventSpec(ops.InstallEvent, 'install')
         hook_path = self.hooks_dir / 'install'
-        for ((rel, ind), path) in {
-                # relative and indirect
-                (True, True): Path('../dispatch'),
-                # relative and direct
-                (True, False): Path(self.charm_exec_path),
-                # absolute and direct
-                (False, False): (self.hooks_dir / self.charm_exec_path).resolve(),
-                # absolute and indirect
-                (False, True): self.JUJU_CHARM_DIR / 'dispatch',
-        }.items():
-            with self.subTest(path=path, rel=rel, ind=ind):
-                # sanity check
-                assert path.is_absolute() == (not rel)
-                assert (path.with_suffix('').name == 'dispatch') == ind
-                try:
-                    hook_path.symlink_to(path)
-
-                    state = self._simulate_event(event)
-                    assert isinstance(state, ops.BoundStoredState)
-
-                    # the .on. was only called once
-                    assert list(state.observed_event_types) == ['InstallEvent']
-                    assert list(state.on_install) == ['InstallEvent']
-                finally:
-                    hook_path.unlink()
-
-    def test_hook_and_dispatch_but_hook_is_dispatch_copy(self):
+
+        path_type_dict = {
+            "indirect": Path('../dispatch'),
+            "direct": Path(self.charm_exec_path),
+            "absolute_direct": (self.hooks_dir / self.charm_exec_path).resolve(),
+            "absolute_indirect": self.JUJU_CHARM_DIR / 'dispatch'
```

Comment:
I tried, but since it has "self", it can't be put in the parametrize, so I put a string in the parametrize then use a dict to resolve to the Path object. Let me do more research on this.

Reply:
Maybe we should avoid parametrize for this, and just do it in a loop (without subtest), or make a helper function and do 4 tests?